File: app/models/event.rb
# frozen_string_literal: true

# == Schema Information
#
# Table name: events
#
#  id             :uuid             not null, primary key
#  properties     :jsonb            not null
#  publish_at     :datetime
#  publish_until  :datetime
#  created_at     :datetime         not null
#  updated_at     :datetime         not null
#  slug           :string
#  location_id    :uuid
#  event_id       :uuid
#  name           :string           not null
#  start_at       :datetime         not null
#  end_at         :datetime         not null
#  website        :string
#  official       :boolean          default(FALSE), not null
#  abstract       :text
#  description    :text
#  ticket_link    :string
#  organizer_name :string
#  categories     :text             default([]), not null, is an Array
#  series         :string
#  full_day       :boolean          default(FALSE), not null
#  bluf_id        :string
#

class Event < ApplicationRecord
  extend FriendlyId
  friendly_id :name, use: :slugged
  include DocumentSerializable
  CATEGORIES = ["bluf", "csd", "culture", "election", "festival", "party", "social"].freeze

  has_many :travel_plans, dependent: :delete_all
  has_many :events, dependent: :destroy
  belongs_to :location

  scope :published, -> { where("events.publish_at <= NOW()") }
  scope :with_attachments, -> { with_attached_hero.with_attached_header.with_attached_logo.with_attached_flyer.with_attached_gallery_images }
  scope :in_future, -> { where("events.end_at >= NOW()") }
  scope :awaiting_review, -> { where(publish_at: nil).in_future }
  scope :chronologic, -> { order(start_at: :asc) }
  scope :happening_in_month, ->(month) { where("events.start_at > ? AND events.start_at < ?", month.beginning_of_month, month.end_of_month) }
  scope :searched, ->(term) { where("events.name ILIKE ?", "%#{term}%") if term.present? }
  scope :listed, -> { published.in_future.chronologic }

  has_many :events, dependent: :destroy
  belongs_to :event, optional: true

  has_and_belongs_to_many :owners, class_name: "User"

  has_one_attached :hero
  has_one_attached :header
  has_one_attached :logo
  has_one_attached :flyer

  has_many_attached :gallery_images

  validates :name, :start_at, :end_at, presence: true

  [:start_at, :end_at].map do |attribute|
    define_method attribute do
      return self[attribute] unless location

      self[attribute].in_time_zone(location.timezone)
    end
  end

  def pending_review?
    publish_at.nil?
  end

  def published?
    publish_at&.past?
  end

  def publish!
    update! publish_at: DateTime.now
  end

  def categories=(values)
    self[:categories] = Array.wrap(values).select { |e| Event::CATEGORIES.include?(e) }
  end

  def to_ics
    event = Icalendar::Event.new
    if full_day
      event.dtstart = Icalendar::Values::Date.new start_at
      event.dtstart.ical_params = { "VALUE" => "DATE" }
      event.dtend = Icalendar::Values::Date.new end_at + 1.day
      event.dtend.ical_params = { "VALUE" => "DATE" }
    else
      event.dtstart = start_at
      event.dtend = end_at
    end
    event.summary = name
    event.description = abstract if abstract.present?
    event.url = "https://#{ENV.fetch('HOST', 'thefetishtraveller.com')}/events/#{slug}"
    event.geo = [location.lat, location.lon] if location&.lat && location&.lon
    event.location = location&.description
    event
  end

  def source_url
    return "https://bluf.com/events/#{bluf_id}" if bluf_id.present?
  end

  def main_event
    event || self
  end
end

File: src/models/event.ts
import moment from 'moment';
import { uniqBy } from 'lodash';

import { DB, State } from '../state';
import { Location } from './location';
import { Syncable } from './syncable';
import { Image } from './image';
import { dateRange, parseDate } from '../util';

export interface Event extends Syncable {
  id: string;
  slug: string;
  eventId?: string;
  series?: string;
  name: string;
  endAt: Date;
  startAt: Date;
  locationId: string;
  abstract?: string;
  description?: string;
  website?: string;
  ticketLink?: string;
  organizerName?: string;
  hero?: Image;
  header?: Image;
  logo?: Image;
  flyer?: Image;
  galleryImages: Image[];
  categories?: string[];
  events?: Event[];
  fullDay: boolean;
  editable?: boolean;
  unsubmitted?: boolean;
}

export const eventCategories = [
  'fetish',
  'csd',
  'election',
  'festival',
  'party',
  'culture',
  'bluf',
  'social',
  'dinner',
  'brunch',
];

export interface EventWithLocation extends Event {
  location?: Location;
  locationSlug?: string;
}

export function formatEventDate(event: {
  startAt: Date | string;
  endAt: Date | string;
  fullDay?: boolean;
  location?: { timezone?: string | null } | null;
}): string {
  return dateRange(
    event.startAt,
    event.endAt,
    event.location ? event.location.timezone : undefined,
    event.fullDay
  );
}

export function joinLocation(
  rawEvents: string[] | Event[],
  state: State
): EventWithLocation[] {
  const db = new DB(state);
  if (rawEvents[0] && typeof rawEvents[0] === 'string') {
    rawEvents = db
      .table('events')
      .where(e => (rawEvents as string[]).includes(e.id));
  }
  const locations = db.table('locations');
  const events: EventWithLocation[] = [];
  (rawEvents as Event[]).forEach(event => {
    const location = locations.find(event.locationId);
    events.push({ ...event, location });
  });
  return events;
}

export function joinSubevents(
  rawEvents: string[] | Event[],
  state: State
): Event[] {
  if (rawEvents[0] && typeof rawEvents[0] === 'string') {
    rawEvents = (rawEvents as string[]).map(e => state.data.events.byId[e]);
  }
  const events: Event[] = [];
  (rawEvents as Event[]).forEach(event => {
    events.push({
      ...event,
      events: events.filter(e => e.eventId === event.id),
    });
  });
  return events;
}

export function isCurrent(event: { endAt: string | Date }): boolean {
  return parseDate(event.endAt) > new Date();
}

export function isRoot(event: { eventId?: string }): boolean {
  return !event.eventId;
}

export function chronological(a: Event, b: Event): number {
  return (a.startAt as any) - (b.startAt as any);
}

export interface Month {
  year: number;
  month: number;
  name: string;
}

export function months(
  events: Array<{ startAt: Date | string }>
): Array<Month> {
  return uniqBy(
    events.map(e => {
      const time = moment(e.startAt);
      return {
        year: time.year(),
        month: time.month(),
        name: time.format('MMM YYYY'),
      };
    }),
    'name'
  );
}

export function inMonth(
  event: { startAt: Date | string },
  month: { year: number; month: number }
) {
  const time = moment(event.startAt);
  return time.month() === month.month && time.year() === month.year;
}

export function byMonth<T extends { startAt: Date | string }>(
  events: T[]
): Array<{ date: Date; events: T[] }> {
  const buckets: { [date: string]: T[] } = {};
  events.forEach(event => {
    const date = moment(event.startAt);
    const dateString = [date.year(), date.month() + 1].join('-');
    buckets[dateString] = buckets[dateString] || [];
    buckets[dateString].push(event);
  });
  const result: Array<{ date: Date; events: T[] }> = [];
  Object.keys(buckets).forEach(key => {
    result.push({
      date: moment(key + '-01').toDate(),
      events: buckets[key],
    });
  });
  return result;
}

export function byMainEvent<T extends { eventId?: string | null; id: string }>(
  events: T[]
): Array<{ event: T; events: T[] }> {
  const buckets: { [eventId: string]: T[] } = {};
  events.forEach(event => {
    const id = event.eventId || event.id;
    buckets[id] = buckets[id] || [];
    buckets[id].push(event);
  });
  const result: Array<{ event: T; events: T[] }> = [];
  Object.keys(buckets).forEach(key => {
    result.push({
      event: buckets[key].filter(e => !e.eventId)[0],
      events: buckets[key].filter(e => e.eventId),
    });
  });
  return result;
}

export function matchesTerm(
  event: {
    name: string;
    categories?: string[] | null;
    location?: {
      name: string;
      city?: string | null;
      countryName?: string | null;
    } | null;
  },
  term: string
): boolean {
  let normalizedEvent = (event.name || '').toLocaleLowerCase();
  if (event.location) {
    normalizedEvent += (
      event.location.name +
      (event.location.city || '') +
      (event.location.countryName || '') +
      (event.categories || []).join()
    ).toLocaleLowerCase();
  }
  const normalizedTerm = term.toLocaleLowerCase();
  return normalizedEvent.includes(normalizedTerm);
}
